### ISSUE TITLE
GridSampler border assumptions 

### ISSUE THREAD

**Issue by y0ast** [pytorch/pytorch#2625]
`torch.nn.functional.GridSampler` makes the assumption that when a grid point is out of the original input frame then the sampled pixel value should be zero:

From `torch/lib/THNN/generic/SpatialGridSamplerBilinear.c`:
```
#define SAFE_GET(input, x, y, n, c, H, W) x >= 0 && x < W && y >=0 \
    && y < H ? THTensor_fastGet4d(input, n, c, y, x) : 0
```

I saw that cudnn and [another cuda implementation](https://github.com/qassemoquab/stnbhwd/blob/master/BilinearSamplerBHWD.cu) do the same.

In our research we have found that using the border value works much better. So the safe_get would implement something like `x = min(W, max(x, 0))` and `y = min(H, max(y, 0))`, instead of just returning zero.

We could simulate that behavior with the current implementation by manually clamping the grid before passing it to the sampler, however this leads to incorrect bilinear interpolation weights. Instead of correctly weighting a pixel by its true distance (even if it's outside the frame), it would weigh it by the distance to the border.

I was considering creating a PR and adding this to the above mentioned C code, however in that case the C implementation would deviate from cudnn, which is probably not ideal. So I am putting it up for discussion first. 

EDIT: @fmassa pointed out that the Tensorflow implementation does use my proposed border behavior: https://github.com/tensorflow/models/blob/master/transformer/spatial_transformer.py#L89

**Comment by fmassa** [comment]
In order to maintain backward compatibility, we could add an option to the backends to alternate between both behaviors, but that would require changes to cudnn, for which we don't have access to. @ngimel is that something that we could add to cudnn?

**Comment by soumith** [comment]
this seems like a decent idea, except that on the CUDA side we'd deviate from CuDNN. Which means that we need a custom CUDA kernel in THCUNN, or we need to write a very slow op in terms of autograd functions (like TF did. I doubt that the TF implementation is fast).
What do you guys prefer? I think the CUDA kernel wont be trivial.

**Comment by y0ast** [comment]
It would be great if CuDNN would support this as an option. Do you know if there is anywhere we can make a feature request? 

I have an implementation in terms of autograd functions, but it's indeed much slower than the GridSampler. The custom CUDA kernel is a last resort option, from my point of view.

**Comment by soumith** [comment]
Someone at Facebook is implementing a THCUNN version of this layer, and then we can easily add a border value option.

**Comment by ngimel** [comment]
I've passed on the request to cudnn.

**Comment by soumith** [comment]
@y0ast now that THCUNN has a hackable implementation that's quite similar to the THNN implementation, feel free to send a PR with adding the `border` padding instead of `zero` padding.

**Comment by y0ast** [comment]
Cool, I'll try to do this next week!

**Comment by soumith** [comment]
fixed via https://github.com/pytorch/pytorch/pull/3599
